modules/codegen/src/main/scala/com/twilio/guardrail/generators/Java/AsyncHttpClientClientGenerator.scala

```diff
@@ -0,0 +1,770 @@
+        }
+        def createBuilderConstructorAssignment(name: String): Statement = {
+          new ExpressionStmt(new AssignExpr(
+            new FieldAccessExpr(new ThisExpr, name),
+            new MethodCallExpr("requireNonNull", new NameExpr(name)), AssignExpr.Operator.ASSIGN)
+          )
+        }
+        (serverUrl, tracingName) match {
+          case (None, None) if tracing =>
+            builderConstructor.setParameters(new NodeList(
+              createConstructorParameter(URI_TYPE, "baseUrl"),
+              createConstructorParameter(STRING_TYPE, "clientName")
+            ))
+            builderConstructor.setBody(new BlockStmt(new NodeList(
+              createBuilderConstructorAssignment("baseUrl"),
+              createBuilderConstructorAssignment("clientName")
+            )))
+
+          case (Some(_), None) if tracing =>
+            builderConstructor.setParameters(new NodeList(
+              createConstructorParameter(STRING_TYPE, "clientName")
+            ))
+            builderConstructor.setBody(new BlockStmt(new NodeList(
+              createBuilderConstructorAssignment("clientName")
+            )))
+
+          case (None, _) =>
+            builderConstructor.setParameters(new NodeList(
+              createConstructorParameter(URI_TYPE, "baseUrl")
+            ))
+            builderConstructor.setBody(new BlockStmt(new NodeList(
+              createBuilderConstructorAssignment("baseUrl")
+            )))
+
+          case (Some(_), Some(_)) =>  // no params
+
+          case (Some(_), _) if !tracing =>  // no params
+        }
+
+        def addSetter(tpe: Type, name: String, initializer: String => Expression): Unit = {
+          val setter = builderClass.addMethod(s"with${name.capitalize}", PUBLIC)
+          setter.setType(BUILDER_TYPE)
+          setter.addParameter(new Parameter(util.EnumSet.of(FINAL), tpe, new SimpleName(name)))
+          setter.setBody(new BlockStmt(new NodeList(
+            new ExpressionStmt(new AssignExpr(new FieldAccessExpr(new ThisExpr, name), initializer(name), AssignExpr.Operator.ASSIGN)),
+            new ReturnStmt(new ThisExpr)
+          )))
+        }
+        val nonNullInitializer: String => Expression = name => new MethodCallExpr(null, "requireNonNull", new NodeList[Expression](new NameExpr(name)))
+        def optionalInitializer(valueArg: String => Expression): String => Expression = name => new MethodCallExpr(new NameExpr("Optional"), "of", new NodeList[Expression](valueArg(name)))
+        if (serverUrl.isDefined) {
+          addSetter(URI_TYPE, "baseUrl", nonNullInitializer)
+        }
+        if (tracing) {
+          addSetter(STRING_TYPE, "clientName", nonNullInitializer)
+        }
+        addSetter(HTTP_CLIENT_FUNCTION_TYPE, "httpClient", optionalInitializer(new NameExpr(_)))
+        addSetter(OBJECT_MAPPER_TYPE, "objectMapper", optionalInitializer(name => new MethodCallExpr(
```

Comment:
Not sure what you're asking...?

Reply:
So, given
```java
        public Builder withObjectMapper(final ObjectMapper objectMapper) {
            this.objectMapper = Optional.of(JacksonSupport.configureObjectMapper(objectMapper));
            return this;
        }
```

I'm wondering about flipping the responsibility so the user has the very last bit of control over the object mapper before being used. Something like

```java
        public Builder withObjectMapper(final Function<ObjectMapper, ObjectMapper> func) {
            this.objectMapper = Optional.of(
                func(JacksonSupport.configureObjectMapper(new ObjectMapper()))
            );
            return this;
        }
```

Reply:
Hmm.  That doesn't allow the caller to pass an existing `ObjectMapper`, though.

Reply:
Sure, you'd just ignore the input and return a static output

Reply:
the flipside is that the user-supplied ObjectMapper would need all the stuff that you're configuring for free here

Reply:
I'm not really sure what this buys us, though.  The current API allows:

1. User supplies their own ObjectMapper, and then we configure it to ensure it can handle at least the types we may make use of.
2. User doesn't care, so we construct one ourselves and again configure it to ensure it can handle at least the types we make use of.

For (2), they're not giving us anything, so... whatevs.

For (1), they're giving us an ObjectMapper they've created (or gotten from somewhere else), so they've already been able to configure it as they've seen fit before giving it to us.  Yes, we then configure it further, but in a way that's required for proper operation.  Anything they do after that (such as remove a module we add) can only break things.

Reply:
Oh, hmm. A design decision in the back of my head has always been to provide workarounds for the safe stuff, in case modeling has failed in some way. I was thinking about what happens when there's a bug in or OM configuration; is someone blocked and has to wait for a bugfix or try to fix it themselves, or can they keep going with degraded functionality? I guess if they still have ownership of the ObjectMapper, this is fine, size they can just alter what we've configured.

Reply:
Yeah, if we screw up, they can still configure the OM after we've broken it.